File: backend/src/routes/announcements.js
// src/routes/announcements.js
const express = require('express');
const router = express.Router();
const db = require('../config/db');
const { authenticateToken, requireRole } = require('../middleware/auth');
const { z } = require('zod');
const { sendEmail } = require('../utils/mailer');
const logger = require('../utils/logger');

// Schema validation for creating an announcement
const announcementSchema = z.object({
  title: z.string().min(1),
  content: z.string().min(1),
  audience_scope: z.enum(['all-students', 'all-mentors', 'team']),
  team_id: z.string().uuid().optional().nullable(),
}).refine(data => {
  if (data.audience_scope === 'team' && !data.team_id) {
    return false;
  }
  return true;
}, {
  message: "team_id is required when audience_scope is 'team'",
  path: ['team_id']
});

/**
 * GET /api/announcements
 * Retrieve announcements filtered by user role and team membership
 */
router.get('/', authenticateToken, async (req, res) => {
  const user_id = req.user.id;
  const role = req.user.role;

  try {
    let queryText = '';
    let queryParams = [];

    if (role === 'admin') {
      // Admins can see all announcements
      queryText = 'SELECT * FROM announcements ORDER BY created_at DESC';
    } else if (role === 'mentor') {
      // Mentors see all-mentors announcements
      queryText = "SELECT * FROM announcements WHERE audience_scope = 'all-mentors' ORDER BY created_at DESC";
    } else {
      // Students see all-students announcements, plus their team specific ones
      // First find user's team if any
      const teamRes = await db.query(
        'SELECT team_id FROM team_members WHERE user_id = $1',
        [user_id]
      );
      
      const team_id = teamRes.rows.length > 0 ? teamRes.rows[0].team_id : null;

      if (team_id) {
        queryText = `
          SELECT * FROM announcements 
          WHERE audience_scope = 'all-students' 
             OR (audience_scope = 'team' AND team_id = $1)
          ORDER BY created_at DESC`;
        queryParams = [team_id];
      } else {
        queryText = `
          SELECT * FROM announcements 
          WHERE audience_scope = 'all-students'
          ORDER BY created_at DESC`;
      }
    }

    const { rows } = await db.query(queryText, queryParams);
    res.json(rows);
  } catch (err) {
    console.error('Fetch announcements error:', err);
    res.status(500).json({ error: 'Failed to fetch announcements due to a server error.' });
  }
});

/**
 * POST /api/announcements
 * Create a new announcement (Admin only)
 */
router.post('/', authenticateToken, requireRole(['admin', 'mentor']), async (req, res) => {
  const validation = announcementSchema.safeParse(req.body);
  if (!validation.success) {
    return res.status(400).json({ error: validation.error.errors });
  }

  const { title, content, audience_scope, team_id } = validation.data;
  const userRole = req.user.role;
  const userId = req.user.id;

  try {
    // Role-based validation and security check
    if (userRole === 'mentor') {
      if (audience_scope !== 'team') {
        return res.status(403).json({ error: 'Access forbidden: Mentors can only create team-specific announcements.' });
      }
      if (!team_id) {
        return res.status(400).json({ error: 'team_id is required for team-specific announcements.' });
      }
      // Check if team exists and belongs to the mentor
      const teamCheck = await db.query('SELECT id, mentor_id FROM teams WHERE id = $1', [team_id]);
      if (teamCheck.rows.length === 0) {
        return res.status(404).json({ error: 'Target team not found.' });
      }
      if (teamCheck.rows[0].mentor_id !== userId) {
        return res.status(403).json({ error: 'You are not authorized to post announcements to this team.' });
      }
    } else if (userRole === 'admin') {
      // If targeted to a specific team, check if team exists
      if (audience_scope === 'team' && team_id) {
        const teamCheck = await db.query('SELECT id FROM teams WHERE id = $1', [team_id]);
        if (teamCheck.rows.length === 0) {
          return res.status(404).json({ error: 'Target team not found.' });
        }
      }
    }

    const { rows } = await db.query(
      `INSERT INTO announcements (title, content, audience_scope, team_id)
       VALUES ($1, $2, $3, $4)
       RETURNING id, title, content, audience_scope, team_id, created_at`,
      [title, content, audience_scope, audience_scope === 'team' ? team_id : null]
    );

    // If target audience is all students, trigger a non-blocking email broadcast
    if (audience_scope === 'all-students') {
      db.query("SELECT email FROM users WHERE role = 'student'")
        .then(async studentRes => {
          const studentEmails = studentRes.rows.map(r => r.email);
          const emailPromises = studentEmails.map(email =>
            sendEmail({
              to: email,
              subject: `New Announcement: ${title}`,
              htmlBody: `
                <h3>New Announcement Posted</h3>
                <p><strong>Title:</strong> ${title}</p>
                <p>${content.replace(/\n/g, '<br/>')}</p>
                <hr/>
                <p>Please check your student dashboard on the portal for updates.</p>
                <p>Best regards,<br/>ITE Startup Launch Pad</p>
              `
            })
          );
          await Promise.allSettled(emailPromises);
        })
        .catch(emailQueryErr => {
          logger.error('Failed to retrieve student emails for announcement broadcast:', emailQueryErr);
        });
    }

    res.status(201).json(rows[0]);
  } catch (err) {
    console.error('Create announcement error:', err);
    res.status(500).json({ error: 'Failed to create announcement due to a server error.' });
  }
});

/**
 * DELETE /api/announcements/:id
 * Delete an announcement (Admin only)
 */
router.delete('/:id', authenticateToken, requireRole('admin'), async (req, res) => {
  const { id } = req.params;
  try {
    const { rowCount } = await db.query('DELETE FROM announcements WHERE id = $1', [id]);
    if (rowCount === 0) return res.status(404).json({ error: 'Announcement not found.' });
    res.json({ message: 'Announcement deleted.' });
  } catch (err) {
    console.error('Delete announcement error:', err);
    res.status(500).json({ error: 'Failed to delete announcement.' });
  }
});

module.exports = router;

File: backend/routers/announcements.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from database import get_db
from models import User, Announcement
from routers.auth import get_current_user
import uuid

router = APIRouter(prefix="/api/announcements", tags=["announcements"])


def ann_to_dict(a: Announcement) -> dict:
    return {
        "id": a.id, "title": a.title, "body": a.body,
        "createdById": a.created_by_id, "createdByRole": a.created_by_role,
        "createdByName": a.created_by_name, "recipients": a.recipients,
        "teamId": a.team_id, "createdAt": a.created_at,
    }


class CreateAnnouncementRequest(BaseModel):
    title: str
    body: Optional[str] = None
    recipients: str = "all"  # "all" | "students" | "mentors" | "team-<id>"
    teamId: Optional[str] = None


from sqlalchemy import or_

@router.get("")
def list_announcements(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    query = db.query(Announcement)

    if current_user.role == "admin":
        # Admin can see all global and targeted announcements
        pass
    elif current_user.role == "mentor":
        # Mentors see global, mentor-specific, and all team announcements
        query = query.filter(
            or_(
                Announcement.recipients == "all",
                Announcement.recipients == "mentors",
                Announcement.recipients.startswith("team-")
            )
        )
    elif current_user.role == "student":
        # Students see global, student-specific, and their own team announcements
        filters = [
            Announcement.recipients == "all",
            Announcement.recipients == "students"
        ]
        if current_user.team_id:
            filters.append(Announcement.recipients == f"team-{current_user.team_id}")
        query = query.filter(or_(*filters))

    # Optimal SQL ordering and fetching
    all_anns = query.order_by(Announcement.created_at.desc()).all()
    
    return [ann_to_dict(a) for a in all_anns]


@router.post("")
def create_announcement(body: CreateAnnouncementRequest, db: Session = Depends(get_db),
                        current_user: User = Depends(get_current_user)):
    if current_user.role not in ("admin", "mentor"):
        raise HTTPException(status_code=403, detail="Admin/Mentor only")
    a = Announcement(
        id=str(uuid.uuid4()),
        title=body.title,
        body=body.body,
        created_by_id=current_user.id,
        created_by_role=current_user.role,
        created_by_name=current_user.name,
        recipients=body.recipients,
        team_id=body.teamId,
    )
    db.add(a)
    db.commit()
    return ann_to_dict(a)


@router.delete("/{ann_id}")
def delete_announcement(ann_id: str, db: Session = Depends(get_db),
                        current_user: User = Depends(get_current_user)):
    if current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Admin only")
    a = db.query(Announcement).filter(Announcement.id == ann_id).first()
    if not a:
        raise HTTPException(status_code=404, detail="Not found")
    db.delete(a)
    db.commit()
    return {"success": True}
